src/Thread.tsx

```diff
@@ -0,0 +1,51 @@
+import {defined} from './Base';
+import {fetchMessageIdsAndLabels, fetchMessagesByIds} from './Gapi';
+import {Message} from './Message';
+
+export class Thread {
+  private _rawThread: gapi.client.gmail.Thread;
+  private _messages: Message[] | undefined;
+  private _firstAndLastMessages: Message[] | undefined;
+
+  constructor(thread: gapi.client.gmail.Thread) {
+    this._rawThread = thread;
+    this._messages = undefined;
+    this._firstAndLastMessages = undefined;
+  }
+
+  id(): string {
+    return defined(this._rawThread.id);
+  }
+
+  messages(): Message[] {
+    return defined(this._messages);
+  }
+
+  firstAndLastMessages(): Message[] | undefined {
+    return this._firstAndLastMessages;
```

Comment:
Should this be defined(this._firstAndLastMessages)? If we're calling this before the messages have been fetched, it's a bug, right?